### ISSUE TITLE
Kotlin Scripting is not possible due to the Quarkus Base Runtime classloader

### ISSUE THREAD

**Issue by ckosmowski** [quarkusio/quarkus#43271]
### Describe the bug

I am trying to use Kotlin scripting inside my quarkus application. I tried in two ways (using kotlins older jsr223 implementation and the newer ScriptingHost approach). I use a script and i am trying to pass objects of a type that is defined in the classpath of my quarkus application.

### Expected behavior

The script is able to execute and load the classes of the objects i pass to it

### Actual behavior

The script fails to load the classes. I debugged it until i found that the contexts class loader is always the Quarkus Base Runtime classLoader which is not able to load classes of my application.

### How to Reproduce?

1. Create a quarkus project
2. Try to execute koltin scripts (either jsr223 oder ScriptingHost)
3. Try to pass an object to the script execution that is of a non primitive type of a class located in the application itsself.

**Comment by ckosmowski** [comment]
I did not know what exactly to pass to the environment i tried tons of jdks, kotlin versions and so on

**Comment by geoand** [comment]
Any chance you can upload a sample that shows what you are trying to do and the problems you are facing?

Thanks

**Comment by ckosmowski** [comment]
Yes, i'm currently working on a super stripped down working example without quarkus. Would that be sufficient?

**Comment by ckosmowski** [comment]
Hi im managed to get both in one project:

https://github.com/ckosmowski/kotlin-script-reproducer

1.) Start the Main class directly -> The script works. It outputs "Hello World!" and "A Model From Outside"
2.) run quarkus dev -> The output will be "Hello World" and a longish output with the message " ERROR Unresolved reference 'printName'. " at its end.

Which according to my research has something to do with classloaders and especially quarkus class loader which is the point where i am completely lost.

If you need anything more, please let me know.

Forgot to say: I Love Quarkus, thanks for the great work.

**Comment by geoand** [comment]
Thanks a lot, I'll try it out and let you know!

**Comment by ckosmowski** [comment]
Something to add, there is a way that you can usually replace the used classloader like this:

```kotlin
val result = host.eval(StringScriptSource("""
            printName()
        """.trimIndent()), ScriptWithMavenDepsConfiguration, ScriptEvaluationConfiguration {
            implicitReceivers(modelFromOutside)
            jvm {
                baseClassLoader(FilteringClassLoader(Thread.currentThread().contextClassLoader, listOf("org.example", "java.lang")))
            }
        })
```

FilterClassLoader:

```kotlin
class FilteringClassLoader(
        parent: ClassLoader,
        val allowedPackages: List<String>
    ) : ClassLoader(parent) {

        override fun loadClass(name: String, resolve: Boolean): Class<*> {
            if (allowedPackages.any { name.startsWith(it) }) {
                return super.loadClass(name, resolve)
            }
            throw ClassNotFoundException("Class $name is not allowed")
        }
    }
```

I hoped that this does the trick but however this works in the plain java version but not with quarkus dev.

**Comment by geoand** [comment]
That's unfortunate...

**Comment by ckosmowski** [comment]
Is there anything i can do to assist?

**Comment by geoand** [comment]
I have not had time to look into it unfortunately.
